File: src/test/java/io/javalin/TestErrorMapper.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin;

import org.junit.jupiter.api.Test;

import static org.hamcrest.MatcherAssert.*;
import static org.hamcrest.Matchers.*;

public class TestErrorMapper extends _UnirestBaseTest {

    @Test
    public void test_404mapper_works() throws Exception {
        app.error(404, ctx -> {
            ctx.result("Custom 404 page");
        });
        assertThat(GET_body("/unmapped"), is("Custom 404 page"));
    }

    @Test
    public void test_500mapper_works() throws Exception {
        app.get("/exception", ctx -> {
            throw new RuntimeException();
        }).error(500, ctx -> {
            ctx.result("Custom 500 page");
        });
        assertThat(GET_body("/exception"), is("Custom 500 page"));
    }

    @Test
    public void testError_higherPriority_thanException() throws Exception {
        app.get("/exception", ctx -> {
            throw new RuntimeException();
        }).exception(Exception.class, (e, ctx) -> {
            ctx.status(500).result("Exception handled!");
        }).error(500, ctx -> {
            ctx.result("Custom 500 page");
        });
        assertThat(GET_body("/exception"), is("Custom 500 page"));
    }

    @Test
    public void testError_throwingException_isCaughtByExceptionMapper() throws Exception {
        app.get("/exception", ctx -> {
            throw new RuntimeException();
        }).exception(Exception.class, (e, ctx) -> {
            ctx.status(500).result("Exception handled!");
        }).error(500, ctx -> {
            ctx.result("Custom 500 page");
            throw new RuntimeException();
        });
        assertThat(GET_body("/exception"), is("Exception handled!"));
    }

}

File: src/test/java/io/javalin/TestErrorMapper.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin

import io.javalin.util.TestUtil
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.junit.Test

class TestErrorMapper {

    @Test
    fun `error-mapper works for 404`() = TestUtil.test { app, http ->
        app.error(404) { ctx -> ctx.result("Custom 404 page") }
        assertThat(http.getBody("/unmapped"), `is`("Custom 404 page"))
    }

    @Test
    fun `error-mapper works for 500`() = TestUtil.test { app, http ->
        app.get("/exception") { ctx -> throw RuntimeException() }
                .error(500) { ctx -> ctx.result("Custom 500 page") }
        assertThat(http.getBody("/exception"), `is`("Custom 500 page"))
    }

    @Test
    fun `error-mapper runs after exception-mapper`() = TestUtil.test { app, http ->
        app.get("/exception") { ctx -> throw RuntimeException() }
                .exception(Exception::class.java) { e, ctx -> ctx.status(500).result("Exception handled!") }
                .error(500) { ctx -> ctx.result("Custom 500 page") }
        assertThat(http.getBody("/exception"), `is`("Custom 500 page"))
    }

    @Test
    fun `error-mapper can throw exceptions`() = TestUtil.test { app, http ->
        app.get("/exception") { ctx -> throw RuntimeException() }
                .exception(Exception::class.java) { e, ctx -> ctx.status(500).result("Exception handled!") }
                .error(500) { ctx ->
                    ctx.result("Custom 500 page")
                    throw RuntimeException()
                }
        assertThat(http.getBody("/exception"), `is`("Exception handled!"))
    }

}
